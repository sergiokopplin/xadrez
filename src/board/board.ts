import { getPositionAxis, Position } from '../utils'

export const whitePieces = 'rnbqkp'
export const blackPieces = 'RNBQKP'

export const defaultRows = ['rnbqkbnr', 'pppppppp', null, null, null, null, 'PPPPPPPP', 'RNBQKBNR']

export interface Square {
  piece: string
}

export type Color = 'black' | 'white'

export class Board {
  public readonly board: Square[]

  constructor () {
    this.board = this.populateBoard()
  }

  private fixArrayIndex (y: number, x: number): number[] {
    return [y - 1, x - 1]
  }

  private populateBoard (): Square[] {
    const board = new Array(8).fill([])
    const emptyRow = new Array(8).fill(null)

    defaultRows.forEach((item, rowIndex) => {
      if (!item) {
        board[rowIndex] = emptyRow
        return
      }

      board[rowIndex] = item.split('').map(piece => ({
        piece
      }))
    })

    return board
  }

  public inferPieceColor (position: Position): Color {
    const [Y, X] = getPositionAxis(position)
    const [y, x] = this.fixArrayIndex(Y, X)
    const piece = this.board[y][x]?.piece

    if (!piece) return null

    if (whitePieces.includes(piece)) {
      return 'white'
    }

    if (blackPieces.includes(piece)) {
      return 'black'
    }
  }

  public isOpponent (ourPosition: Position, theirPosition: Position): boolean {
    const ourColor = this.inferPieceColor(ourPosition)
    const theirColor = this.inferPieceColor(theirPosition)

    if (!ourColor || !theirColor) {
      return false
    }
  }

  public getSquare (position: Position): Square {
    const [Y, X] = getPositionAxis(position)
    const [y, x] = this.fixArrayIndex(Y, X)

    return this.board[y][x]
  }
}
