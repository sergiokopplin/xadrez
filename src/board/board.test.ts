import { Board, defaultRows } from './board'

describe('Board', () => {
  test('Should Create a board', () => {
    expect(new Board()).toBeTruthy()
  })

  test('Should Create a board and populate squares correctly', () => {
    const emptyRow = new Array(8).fill(null)
    const sut = new Board()

    expect(sut.board).toEqual(defaultRows.map(item => item || emptyRow))
  })
})
