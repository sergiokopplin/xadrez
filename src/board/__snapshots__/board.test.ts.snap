// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board constructor ensure board create and populate squares correctly 1`] = `
Array [
  Array [
    Object {
      "piece": "r",
    },
    Object {
      "piece": "n",
    },
    Object {
      "piece": "b",
    },
    Object {
      "piece": "q",
    },
    Object {
      "piece": "k",
    },
    Object {
      "piece": "b",
    },
    Object {
      "piece": "n",
    },
    Object {
      "piece": "r",
    },
  ],
  Array [
    Object {
      "piece": "p",
    },
    Object {
      "piece": "p",
    },
    Object {
      "piece": "p",
    },
    Object {
      "piece": "p",
    },
    Object {
      "piece": "p",
    },
    Object {
      "piece": "p",
    },
    Object {
      "piece": "p",
    },
    Object {
      "piece": "p",
    },
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
  ],
  Array [
    Object {
      "piece": "P",
    },
    Object {
      "piece": "P",
    },
    Object {
      "piece": "P",
    },
    Object {
      "piece": "P",
    },
    Object {
      "piece": "P",
    },
    Object {
      "piece": "P",
    },
    Object {
      "piece": "P",
    },
    Object {
      "piece": "P",
    },
  ],
  Array [
    Object {
      "piece": "R",
    },
    Object {
      "piece": "N",
    },
    Object {
      "piece": "B",
    },
    Object {
      "piece": "Q",
    },
    Object {
      "piece": "K",
    },
    Object {
      "piece": "B",
    },
    Object {
      "piece": "N",
    },
    Object {
      "piece": "R",
    },
  ],
]
`;

exports[`Board customRows ensure board creation with custom rows 1`] = `
"........
........
........
........
ppppPPPP
........
........
........
"
`;

exports[`Board printBoard ensure board print 1`] = `
"RNBQKBNR
PPPPPPPP
........
........
........
........
pppppppp
rnbqkbnr
"
`;
